stackrox/stackrox
# ROX-10665: required attributes auth provider

## Description

Within Managed Services we want to set required attributes returned by auth providers, so that we can ensure only users from the same Red Hat organization are allowed to login. Other Red Hat users shouldn't be issued a Rox token.

For the solution itself, initially we wanted to opt into a non-API exposed implementation, but since we want to support customers changing the auth provider, which in itself triggers a re-creation of it, I had to add additional fields to `storage.AuthProvider`.
The PR does not _yet_ allow for required attributes for _all_ auth providers, but instead limits modifications to required attributes. I'm happy to less strict if we decide its fine for us.

Right now, the following has been done as a summary:
- Create a new field `RequiredAttributes` on the `storage.AuthProvider`.
- Create a new interface `AttributeChecker`, which is added to the `Provider` interface.
- Create an implementation of `AttributeChecker` called `RequiredAttributesChecker`, which will take a list of `RequiredAttribute` and verify whether the user has the attributes set correctly.
- Add a `ProviderOption` that will allow us to inject the `AttributeChecker` during `AuthProvider`.
- Disallow setting / modifications to `RequiredAttribute` from API, as we want to limit changes being done only by the injected configuration from within managed services ([here is the ADR that goes into detail which mechanism we will use for this](https://docs.google.com/document/d/1piqcfXd_H9jVuFxd0vaqX4wEo2c3nOMdoEnGRQ4XCGM/edit#heading=h.h06w8h600t3e))

What still has to be done after this PR has been merged:
- Once the ADR and mechanism is implemented to receive the auth provider configuration for RH SSO provider in managed service context, the auth provider needs to be created with `ProviderOption`.
- Wire modification within the UI to make changes to required attributes so we ship required attributes as a new feature for all auth providers.
- Add E2E tests going through the flow of adding an auth provider with required attributes, verifying that login is successful / unsuccessful with matching / non-matching attributes.


Once an auth provider has been created with required attributes, each required attribute will be listed:
![Screenshot 2022-05-19 at 13 11 31](https://user-images.githubusercontent.com/89904305/169280382-75c7885d-4500-414e-856d-363f0caeb00e.png)

If a user fails to provide the required attributes, he will be rejected with the following error message:
![Screenshot 2022-05-19 at 13 14 17](https://user-images.githubusercontent.com/89904305/169280811-de5361f4-2ed4-4311-bdb1-7a8563936a7d.png)

Overview of the auth provider page:
![Screenshot 2022-05-19 at 13 12 11](https://user-images.githubusercontent.com/89904305/169280470-eee62786-b2ee-4c85-847f-429d818ea9ac.png)

## Checklist

If any of these don't apply, please comment below.

## Testing Performed
- Unit tests
- Manual testing:
  - Create auth provider (i.e. openshift auth / auth0) via API, set required attributes (i.e. name).
Sample auth provider via Auth0:
```json
{
    "name": "Test required attributes",
    "type": "auth0",
    "uiEndpoint": "localhost:8000",
    "enabled": true,
    "active": true,
    "config": {
        "client_id": "<redacted>",
        "issuer": "<redacted>",
        "mode": "fragment"
    },
    "requiredAttributes": [
        {
            "attributeKey": "name",
            "attributeValue": "<username>"
        }
    ]
}
```
Sample auth provider for OpenShift Auth:
```json
{
    "name": "Test req attributes",
    "type": "openshift",
    "config": {},
    "uiEndpoint": "localhost:8000",
    "enabled": true,
    "active": true,
    "requiredAttributes": [
        {
            "attributeKey": "name",
            "attributeValue": "haus"
        }
    ]
}
```
  - Make sure to explicitly set the minimum access role, otherwise you could be presented with the error `no valid role, please contact your system administrator`.
  - Ensure users with non-matching attributes will be denied login.
  - Ensure users with matching attributes will be allowed login and successfully assigned the minimum role (if no other access role exists).
  - Ensure active auth providers cannot be changed via PutAuthProviders.
